pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
-				return reconcile.Result{}, err
+				return err
 			}
 		}
-		return reconcile.Result{}, err
-	}
-	externalClusterProvider, err := kubernetesprovider.NewExternalClusterProvider(r.ImpersonationClient, r.Client)
-	if err != nil {
-		return reconcile.Result{}, err
+		return err
 	}
+	return nil
+}
+
+func (r *reconciler) upgradeAction(ctx context.Context,
+	log *zap.SugaredLogger,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider *kubernetesprovider.ExternalClusterProvider,
+	currentManifest []byte,
+	clusterStatus kubermaticv1.Status,
+) (*corev1.Pod, error) {
+	upgradePod := &corev1.Pod{}
 	version, err := externalClusterProvider.GetVersion(ctx, externalCluster)
 	if err != nil {
-		return reconcile.Result{}, err
+		return nil, err
 	}
 	currentVersion := version.String()
+	wantedVersion, err := getWantedVersion(currentManifest)
+	if err != nil {
+		return nil, err
+	}
+	if clusterStatus != kubermaticv1.StatusReconciling && clusterStatus != kubermaticv1.StatusError {
+		if currentVersion != wantedVersion {
+			err = r.Get(ctx, ctrlruntimeclient.ObjectKey{
+				Name:      KubeOneUpgradePod,
+				Namespace: kubernetesprovider.GetKubeOneNamespaceName(externalCluster.Name),
+			},
+				upgradePod,
+			)
+			if err != nil {
+				if !kerrors.IsNotFound(err) {
+					return nil, err
+				}
+			} else {
+				if err := r.Delete(ctx, upgradePod); err != nil {
+					return nil, err
+				}
+			}
+			log.Debugw("Upgrading kubeone cluster", "from", currentVersion, "to", wantedVersion)
+			if upgradePod, err = r.upgradeCluster(ctx, log, externalCluster); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return upgradePod, nil
+}
 
-	manifestSecret := &corev1.Secret{}
-	if err := r.Get(ctx, types.NamespacedName{Namespace: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Namespace, Name: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Name}, manifestSecret); err != nil {
-		return reconcile.Result{}, fmt.Errorf("can not retrieve kubeone manifest secret: %w", err)
+func (r *reconciler) migrateAction(ctx context.Context,
+	log *zap.SugaredLogger,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider *kubernetesprovider.ExternalClusterProvider,
+	currentManifest []byte,
+	clusterStatus kubermaticv1.Status,
+) (*corev1.Pod, error) {
+	currentContainerRuntime, err := externalcluster.CheckContainerRuntime(ctx, externalCluster, externalClusterProvider)
+	if err != nil {
+		return nil, err
 	}
-	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+	wantedContainerRuntime, err := returnWantedContainerRuntime(currentManifest)
+	if err != nil {
+		return nil, err
+	}
+	pod := &corev1.Pod{}
+	if clusterStatus != kubermaticv1.StatusReconciling && clusterStatus != kubermaticv1.StatusError {
+		if currentContainerRuntime != wantedContainerRuntime {
+			if wantedContainerRuntime != "" && wantedContainerRuntime == resources.ContainerRuntimeContainerd {
```

Comment:
Flipping these conditions around can also make this function more readable and keep the code on the left side of the screen.